xarray/core/indexing.py

```diff
@@ -643,7 +665,16 @@ def shape(self) -> tuple[int, ...]:
         return np.broadcast(*self.key.tuple).shape
 
     def get_duck_array(self):
-        array = self.array[self.key]
+
+        if isinstance(self.array, ExplicitlyIndexedNDArrayMixin):
+            if isinstance(self.key, VectorizedIndexer):
+                array = self.array.vindex[self.key]
+            elif isinstance(self.key, OuterIndexer):
+                array = self.array.oindex[self.key]
+            else:
+                array = self.array[self.key]
+        else:
+            array = self.array[self.key]
```

Comment:
same request for a comment here. I know this is a bit messy (see comment in next line) so it would be good to write something to help us in the future :)